clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -128,23 +137,99 @@ record_filter_t::parallel_shard_error(void *shard_data)
-        if (!filters_[i]->parallel_shard_filter(entry, filter_shard_data_[i]))
+        bool stop = false;
+        if (!filters_[i]->parallel_shard_filter(entry, per_shard->filter_shard_data[i],
+                                                &stop))
             output = false;
+        if (stop)
+            per_shard->filter_stop[i] = true;
+        all_stop = all_stop && per_shard->filter_stop[i];
+    }
+
+    // Everything before the timestamp marker in the first unit header is part of the
+    // trace shard's header.
+    if (entry.type == TRACE_TYPE_MARKER && entry.size == TRACE_MARKER_TYPE_TIMESTAMP) {
+        per_shard->in_shard_header = false;
+    }
+
+    // We output everything in the trace shard's header.
+    if (per_shard->in_shard_header) {
+        output = true;
+    }
+
+    // Keep track of the last omitted unit header so that we can output it if and when
+    // we output a trace_entry_t from this unit.
+    if (!output && entry.type == TRACE_TYPE_MARKER) {
+        switch (entry.size) {
+        case TRACE_MARKER_TYPE_TIMESTAMP:
+            // No need to remember the previous unit's header anymore. We're in the
+            // next unit now.
+            per_shard->last_filtered_unit_header.clear();
+            ANNOTATE_FALLTHROUGH;
+        case TRACE_MARKER_TYPE_WINDOW_ID:
+        case TRACE_MARKER_TYPE_CPU_ID:
+            per_shard->last_filtered_unit_header.push_back(entry);
+        }
+    }
+
+    // Since we're outputing something from this unit, output its header.
+    if (output) {
+        for (trace_entry_t &unit_header_entry : per_shard->last_filtered_unit_header) {
+            if (!write_trace_entry(per_shard, unit_header_entry))
+                return false;
+        }
+        per_shard->last_filtered_unit_header.clear();
+    }
+
+    if (!output) {
+        if (entry.type == TRACE_TYPE_FOOTER ||
+            (entry.type == TRACE_TYPE_MARKER &&
+             entry.size == TRACE_MARKER_TYPE_CHUNK_FOOTER)) {
+            output = true;
+        } else if (!all_stop && entry.type == TRACE_TYPE_MARKER) {
+            // Markers that may be required for future instr/memrefs, therefore
+            // must be retained if there's a chance we'll output more entries
+            // later. These are not really tied to a unit header, so to save
+            // space we do not output their unit's header unless some other
+            // instr/memref is output from the same unit. We decided against
+            // buffering them since there may be too many of these.
+            // XXX: This list needs to be kept in sync with our trace format.
+            switch (entry.size) {
+            case TRACE_MARKER_TYPE_PHYSICAL_ADDRESS:
+            case TRACE_MARKER_TYPE_PHYSICAL_ADDRESS_NOT_AVAILABLE:
+            case TRACE_MARKER_TYPE_VIRTUAL_ADDRESS:
+            case TRACE_MARKER_TYPE_CHUNK_INSTR_COUNT:
+                // TODO i#5675: The value of the TRACE_MARKER_TYPE_CHUNK_INSTR_COUNT
```

Comment:
This one seems like what we *should* implement here: this is worth the infrastructure auto-fixing.  I am confused about the value of the other fixups here.